proxy/src/main/java/com/velocitypowered/proxy/plugin/VelocityPluginManager.java

```diff
@@ -105,8 +115,76 @@ public void loadPlugins(Path directory) throws IOException {
       return;
     }
 
-    List<PluginDescription> sortedPlugins = PluginDependencyUtils.sortCandidates(found);
+    // update plugins from update folder before checking for
+    // dependencies and making sorted dependency list
+    if (applyUpdates) {
+      checkNotNull(updateDirectory, "updateDirectory");
+      checkArgument(Files.isDirectory(updateDirectory), "provided update path isn't a directory");
+      checkNotNull(outdatedPluginDirectory, "outdatedPluginDirectory");
+      checkArgument(Files.isDirectory(outdatedPluginDirectory),
+              "provided outdated plugin path isn't a directory");
+      List<PluginDescription> updatesToApply = new ArrayList<>();
+      JavaPluginLoader updateLoader = new JavaPluginLoader(server, updateDirectory);
+      try (DirectoryStream<Path> stream = Files.newDirectoryStream(updateDirectory,
+              p -> p.toFile().isFile() && p.toString().endsWith(".jar"))) {
+        for (Path path : stream) {
+          try {
+            updatesToApply.add(updateLoader.loadCandidate(path));
+          } catch (Exception e) {
+            logger.error("Unable to load plugin candidate {}", path, e);
+          }
+        }
+      }
+
+      // match the to update plugin's id against already
+      // loaded plugins and replace them if match found
+      for (PluginDescription updatedDescription : updatesToApply) {
+        PluginDescription possibleMatch = found.get(updatedDescription.getId());
+        if (updatedDescription.getSource().isEmpty()) { //should not happen but just in case
+          logger.warn("No source found for plugin {} found", updatedDescription.getId());
+          continue;
+        }
+        Path oldPluginPath = null;
+        if (possibleMatch != null) {
+          if (possibleMatch.getSource().isEmpty()) {
+            logger.warn("No source for plugin {} found, continuing without update.",
+                    possibleMatch.getId());
+            continue;
+          }
+          //move old plugin to outdated plugin directory to rollback in case of failure
+          oldPluginPath = possibleMatch.getSource().get();
+          Files.move(oldPluginPath, outdatedPluginDirectory.resolve(oldPluginPath.getFileName()));
```

Comment:
This can fail due to missing file permissions. It should be in a try-catch block.

Another issue with this move call: It doesn't overwrite existing files:
```
[23:05:24] [main/INFO] [com.velocitypowered.proxy.VelocityServer]: Loading plugins...
[23:05:24] [main/ERROR] [com.velocitypowered.proxy.VelocityServer]: Couldn't load plugins
java.nio.file.FileAlreadyExistsException: plugins/outdated/Punishments-Velocity.jar
	at sun.nio.fs.UnixCopyFile.move(UnixCopyFile.java:449) ~[?:?]
	at sun.nio.fs.UnixFileSystemProvider.move(UnixFileSystemProvider.java:266) ~[?:?]
	at java.nio.file.Files.move(Files.java:1432) ~[?:?]
	at com.velocitypowered.proxy.plugin.VelocityPluginManager.loadPlugins(VelocityPluginManager.java:156) ~[velocity-3.2.0-SNAPSHOT-2023-04-06.jar:3.2.0-SNAPSHOT (git-6c196c5c)]
	at com.velocitypowered.proxy.VelocityServer.loadPlugins(VelocityServer.java:368) ~[velocity-3.2.0-SNAPSHOT-2023-04-06.jar:3.2.0-SNAPSHOT (git-6c196c5c)]
	at com.velocitypowered.proxy.VelocityServer.start(VelocityServer.java:227) ~[velocity-3.2.0-SNAPSHOT-2023-04-06.jar:3.2.0-SNAPSHOT (git-6c196c5c)]
	at com.velocitypowered.proxy.Velocity.main(Velocity.java:69) ~[velocity-3.2.0-SNAPSHOT-2023-04-06.jar:3.2.0-SNAPSHOT (git-6c196c5c)]
[23:05:24] [main/INFO] [com.velocitypowered.proxy.VelocityServer]: Loaded 0 plugins
```
Using this should fix that issue:
```suggestion
          Files.move(oldPluginPath, outdatedPluginDirectory.resolve(oldPluginPath.getFileName()), java.nio.file.StandardCopyOption.REPLACE_EXISTING);
```